sumpy/fmm.py

```diff
@@ -639,10 +689,42 @@ def multipole_to_local(self,
             level_start_target_box_nrs,
             target_boxes, src_box_starts, src_box_lists,
             mpole_exps):
+
+        if self.use_fft:
+            pp_mpole_exps = self.pp_multipole_expansion_zeros()
+            events = []
+            for lev in range(self.tree.nlevels):
+                order = self.level_orders[lev]
+                preprocess_kernel = \
+                    self.code.m2l_preprocess_kernel(order, order)
+
+                source_level_start_ibox, source_mpoles_view = \
+                        self.multipole_expansions_view(mpole_exps, lev)
+
+                _, pp_source_mpoles_view = \
+                        self.pp_multipole_expansions_view(pp_mpole_exps, lev)
+
+                tr_classes = self.m2l_translation_class_level_start_box_nrs()
+                # There's no M2L happening in this level
+                if tr_classes[lev] == tr_classes[lev + 1]:
+                    continue
+
+                evt, _ = preprocess_kernel(
+                    self.queue,
+                    src_expansions=source_mpoles_view,
+                    pp_src_expansions=pp_source_mpoles_view,
+                    src_rscale=level_to_rscale(self.tree, lev),
+                    **self.kernel_extra_kwargs
+                )
+                events.append(evt)
+            mpole_exps = pp_mpole_exps
+            mpole_exps_view_func = self.pp_multipole_expansions_view
+        else:
+            mpole_exps_view_func = self.multipole_expansions_view
+
         local_exps = self.local_expansion_zeros()
 
         events = []
```

Comment:
*cough* event management

Reply:
Done in 467e43b6bd8fac000e4c8f0cafc2cdf76f85122a

Reply:
This isn't right yet. I'll fix it up.